File: 1erdepartamental/c++/11.matrices.cpp
// ============================================
// PROGRAMA: Operaciones con Matrices
// DESCRIPCIÓN: Multiplicación escalar y matricial
// ============================================

#include <iostream>
#include <vector>
#include <stdexcept>
#include <iomanip>
using namespace std;


/**
 * Matriz con operaciones sobrecargadas
 * 
 * Demuestra operator* para escalar y matrices
 */
class Matriz {
private:
    vector<vector<double>> __datos;
    size_t __filas;
    size_t __columnas;
    
    void validarDimensiones(size_t filas, size_t columnas) const {
        if (filas == 0 || columnas == 0) {
            throw invalid_argument("Dimensiones deben ser > 0");
        }
    }

public:
    Matriz(size_t filas, size_t columnas) 
        : __filas(filas), __columnas(columnas) {
        validarDimensiones(filas, columnas);
        __datos.resize(filas, vector<double>(columnas, 0));
    }
    
    size_t getFilas() const { return __filas; }
    size_t getColumnas() const { return __columnas; }
    
    void set(size_t i, size_t j, double valor) {
        if (i >= __filas || j >= __columnas) {
            throw out_of_range("Índice fuera de rango");
        }
        __datos[i][j] = valor;
    }
    
    double get(size_t i, size_t j) const {
        if (i >= __filas || j >= __columnas) {
            throw out_of_range("Índice fuera de rango");
        }
        return __datos[i][j];
    }
    
    void llenar(double valor) {
        for (auto& fila : __datos) {
            for (auto& elem : fila) {
                elem = valor;
            }
        }
    }
    
    // Multiplicación escalar
    Matriz operator*(double escalar) const {
        Matriz resultado(__filas, __columnas);
        for (size_t i = 0; i < __filas; i++) {
            for (size_t j = 0; j < __columnas; j++) {
                resultado.set(i, j, __datos[i][j] * escalar);
            }
        }
        return resultado;
    }
    
    // Multiplicación matricial
    Matriz operator*(const Matriz& otra) const {
        if (__columnas != otra.__filas) {
            throw invalid_argument("Dimensiones incompatibles para multiplicación");
        }
        
        Matriz resultado(__filas, otra.__columnas);
        for (size_t i = 0; i < __filas; i++) {
            for (size_t j = 0; j < otra.__columnas; j++) {
                double suma = 0;
                for (size_t k = 0; k < __columnas; k++) {
                    suma += __datos[i][k] * otra.__datos[k][j];
                }
                resultado.set(i, j, suma);
            }
        }
        return resultado;
    }
    
    friend ostream& operator<<(ostream& os, const Matriz& m) {
        for (size_t i = 0; i < m.__filas; i++) {
            os << "[ ";
            for (size_t j = 0; j < m.__columnas; j++) {
                os << setw(6) << fixed << setprecision(2) << m.__datos[i][j];
                if (j < m.__columnas - 1) os << ", ";
            }
            os << " ]" << endl;
        }
        return os;
    }
};


// ============================================
// EJECUCIÓN
// ============================================
int main() {
    Matriz m1(2, 2);
    m1.set(0, 0, 1); m1.set(0, 1, 2);
    m1.set(1, 0, 3); m1.set(1, 1, 4);
    
    cout << "Matriz original:" << endl << m1 << endl;
    
    Matriz m2 = m1 * 2;
    cout << "Multiplicación escalar (x2):" << endl << m2 << endl;
    
    Matriz m3 = m1 * m1;
    cout << "Multiplicación matricial:" << endl << m3 << endl;
    
    return 0;
}

File: 1erdepartamental/python/11.matrices.py
class Matriz:
    def __init__(self, filas, columnas, datos=None):
        self.filas = filas
        self.columnas = columnas
        if datos:
            self.datos = datos
        else:
            self.datos = [[0] * columnas for _ in range(filas)]

    def llenar(self, valores):
        idx = 0
        for i in range(self.filas):
            for j in range(self.columnas):
                if idx < len(valores):
                    self.datos[i][j] = valores[idx]
                    idx += 1

    def __mul__(self, otro):
        if isinstance(otro, (int, float)):
            r = Matriz(self.filas, self.columnas)
            for i in range(self.filas):
                for j in range(self.columnas):
                    r.datos[i][j] = self.datos[i][j] * otro
            return r
        elif isinstance(otro, Matriz):
            r = Matriz(self.filas, otro.columnas)
            for i in range(self.filas):
                for j in range(otro.columnas):
                    s = 0
                    for k in range(self.columnas):
                        s += self.datos[i][k] * otro.datos[k][j]
                    r.datos[i][j] = s
            return r

    def mostrar(self):
        for fila in self.datos:
            print([round(x, 2) for x in fila])


if __name__ == "__main__":
    A = Matriz(2, 3)
    A.llenar([1, 2, 3, 4, 5, 6])
    print("A:")
    A.mostrar()

    print("A * 2:")
    (A * 2).mostrar()

    B = Matriz(3, 2)
    B.llenar([7, 8, 9, 10, 11, 12])
    print("B:")
    B.mostrar()

    print("A * B:")
    (A * B).mostrar()
